src/antidote_crdt_orset.erl

```diff
@@ -82,115 +76,88 @@
+        sets:from_list(Elems)
+    ),
+    {ok, DownstreamOps};
+downstream({remove, Elem}, ORSet) ->
+    ToRemove = get_supporting_tokens(Elem, ORSet),
+    DownstreamOp = {Elem, [], ToRemove},
+    {ok, [DownstreamOp]};
+downstream({remove_all, Elems}, ORSet) ->
+    DownstreamOps = sets:fold(
+        fun(Elem, Ops) ->
+            {ok, [Op]} = downstream({remove, Elem}, ORSet),
+            [Op | Ops]
+        end,
+        [],
+        sets:from_list(Elems)
+    ),
+    {ok, DownstreamOps};
+downstream({reset, {}}, ORSet) ->
     % reset is like removing all elements
-    downstream({remove_all, value(ORDict)}, ORDict).
+    downstream({remove_all, value(ORSet)}, ORSet).
 
 %% @doc apply downstream operations and update an `orset()'.
-%% The first parameter denotes this operation is for adding or removing elements.
-%% For add or add_all, the second element of the tuple is a list of elements and tokens to add
-%% For remove or remove_all, the second element of the tuple is a list of elements and their tokens
-%% to remove.
-%% For update, the second element of the tuple is a list of updates to apply, each of which can
-%% either be add, add_all or remove, remove_all.
--spec update(orset_op(), orset()) ->
-                    {ok, orset()}.
-update({add, {Elem, [Token|_]}}, ORDict) ->
-    add_elem(Elem, Token, ORDict);
-update({add_all, Elems}, ORDict0) ->
-    OD = lists:foldl(fun(Elem, ORDict) ->
-                             {ok, ORDict1} = update({add, Elem}, ORDict),
-                             ORDict1
-                     end, ORDict0, Elems),
-    {ok, OD};
-update({remove, Elem}, ORDict) ->
-    remove_elem(Elem, ORDict);
-update({remove_all, Elems}, ORDict0) ->
-    remove_elems(Elems, ORDict0).
-
--spec equal(orset(), orset()) -> boolean().
-equal(ORDictA, ORDictB) ->
-    ORDictA == ORDictB. % Everything inside is ordered, so this should work
-
-%% @doc the precondition context is a fragment of the CRDT that
-%% operations with pre-conditions can be applied too.  In the case of
-%% OR-Sets this is the set of adds observed.  The system can then
-%% apply a remove to this context and merge it with a replica.
-%% Especially useful for hybrid op/state systems where the context of
-%% an operation is needed at a replica without sending the entire
-%% state to the client.
--spec precondition_context(orset()) -> orset().
-precondition_context(ORDict) ->
-    orddict:fold(fun(Elem, Tokens, ORDict1) ->
-            case minimum_tokens(Tokens) of
-                []      -> ORDict1;
-                Tokens1 -> orddict:store(Elem, Tokens1, ORDict1)
+%% The downstream op is a list of triples.
+%% In each triple:
+%%  - the first component is the elem that was added or removed
+%%  - the second component is the list of supporting tokens to be added
+%%  - the third component is the list of supporting tokens to be removed
+-spec update(downstream_op(), orset()) -> {ok, orset()}.
+update(DownstreamOps, ORSet0) ->
+    ORSet = lists:foldl(
```

Comment:
Similar comment as above: This seems to be quadratic running time. If the elements in the downstream operation are sorted it could be done in linear time.